File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        final = []
        n = len(candidates)
        candidates.sort()
        used = dict()
        def combine(i):
            if sum(res) == target:
                final.append(res[:])
                return
            elif sum(res) > target:
                return
            for j in range(i,n):
                if j == 0  or (candidates[j-1] != candidates[j]) or j-1 in used:
                    used[j] = 1
                    res.append(candidates[j])
                    combine(j+1)
                    res.pop()
                    del used[j]
        combine(0)
        return final
        

File: 40-combination-sum-ii/40-combination-sum-ii.java
class Solution {
    private List<List<Integer>> list = new LinkedList<>();
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        Set<Integer> visited = new HashSet<>();
        combination(candidates, 0, 0, new LinkedList(), target, visited);
        return list;
    }
    public void combination(int[] candidates, int start, int curSum, LinkedList<Integer> curRes, int target, Set visited){
        if(curSum > target)return;
        if(curSum == target){
            list.add((List) curRes.clone());
            
        }
        
        for(int i = start; i < candidates.length; i++){
            if(i > start && candidates[i-1] == candidates[i]&&!visited.contains(i-1))continue;
            curRes.add(candidates[i]);
            visited.add(i);
            combination(candidates, i + 1, curSum + candidates[i], curRes, target, visited);
            curRes.pollLast();
            visited.remove(i);
        }
        
    }
}